Add a book binder the player can use to copy recipes from books

Saicchi opened the pull request:
#### Summary

SUMMARY: Content "Add book binder to copy recipes"

#### Purpose of change

This allows the player to copy only some recipes instead of having to carry an entire book.

#### Describe the solution

The book binder works like a laptop or a tablet. Just like the elink tablet can store recipes from a memory card, with a pen and some paper the player can copy recipes from books you can understand.

#### Describe alternatives you've considered

None.

#### Testing

Runs without error and everything seems to work fine.

#### Additional context

Paper needed for recipe is `1 + difficulty / 2`.
It's capacity is balanced around 15 recipes with difficulty level 10.
To add this book binder it was needed to change paper type from comestible to ammo, fix each sheet of paper having a volume of 0.25L and change paper spawn count to how it was before I implemented the change.
I was considering an option for the player be able to remove recipes once they were added, but I was unsure if I should add it or not.
The recipe for creating a book binder definitely needs to be changed, but I didn't really know what to put there, so it's more like two pieces of plastic with some duct tape and strings binding them and the paper together.


![image](https://user-images.githubusercontent.com/47158232/94188658-18966580-fe80-11ea-9adf-e1053460b023.png)
![image](https://user-images.githubusercontent.com/47158232/94188712-2946db80-fe80-11ea-8ff9-a90801e81af4.png)

Same as a laptop with recipes inside.
![image](https://user-images.githubusercontent.com/47158232/94189470-27c9e300-fe81-11ea-97b4-33244d9c8c40.png)

LaVeyanFiend commented on data/json/itemgroups/Locations_MapExtras/locations.json:
```diff
@@ -347,7 +347,7 @@
       [ "eyedrops", 10 ],
       [ "rag", 30 ],
-      [ "paper", 30 ],
+      { "item": "paper", "prob": 30, "count": 50 },
```
```suggestion
      { "item": "fork", "count": [ 25, 75 ], "prob": 30 },
```
Count I made is arbitrary, but randomizing the amount you find would be appreciated & realistic.

anothersimulacrum reviewed (changes requested):
This should be assigning an activity with duration based on how many pages are used.

Saicchi commented:
I changed it to be a light exercise activity that takes 2 minutes per page. I tested it and didn't have any problems with it. It was my first time adding an activity so please tell me if something is wrong.

anothersimulacrum commented on src/activity_actor.cpp:
```diff
@@ -2139,13 +2140,77 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
+
+void bookbinder_copy_activity_actor::finish( player_activity &act, Character &p )
+{
+    item *it_bkbinder = book_binder.get_item();
```
It's possible that this item location can be invalidated, and you end up with a nullptr here.
```suggestion
    item *it_bkbinder = book_binder.get_item();
    if( it_bkbinder == nullptr ) {
        debugmsg( "Invalid bookbinder for copying recipes" );
        act.set_to_null();
        return;
    }
```

anothersimulacrum commented on src/activity_actor.cpp:
```diff
@@ -2139,13 +2140,77 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
+{
+    item *it_bkbinder = book_binder.get_item();
+
+    const auto old_recipes = it_bkbinder->get_var( "EIPC_RECIPES" );
```
Please replace auto with the proper type.

anothersimulacrum commented on src/activity_actor.cpp:
```diff
@@ -2139,13 +2140,77 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
+
+    const auto old_recipes = it_bkbinder->get_var( "EIPC_RECIPES" );
+    if( old_recipes.empty() ) {
+        it_bkbinder->set_var( "EIPC_RECIPES", "," + r_id + "," );
```
Forgive me for being unfamiliar with this code, but it's supposed to look like `,recipe1,,recipe2,,recipe3,`?

Saicchi replied:
I think it's just as you described.
I copied it from the way the `einkpc_download_memory_card` function stores its recipes. I didn't change the way it was stored to avoid any problems when `item::get_available_recipes` loads the recipe.

anothersimulacrum replied:
That seems odd, but nothing for you to deal with. You could consider adding a function or two to `item` to move the business logic of storing and checking the recipes elsewhere to make this a little cleaner and not duplicate code.

anothersimulacrum commented on src/activity_actor.h:
```diff
@@ -845,6 +845,41 @@ class stash_activity_actor: public activity_actor
+        bookbinder_copy_activity_actor(
+            int pages,
+            const item_location &book_binder,
+            std::string r_id = "",
+            std::string r_name = ""
```
Unless I'm missing something, these should be const references.
```suggestion
            const std::string &r_id = "",
            const std::string &r_name = ""
```

anothersimulacrum commented on src/activity_actor.cpp:
```diff
@@ -2139,13 +2140,77 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
+void bookbinder_copy_activity_actor::start( player_activity &act, Character & )
+{
+    act.moves_total = to_moves<int>( pages * 2_minutes );
+    act.moves_left = to_moves<int>( pages * 2_minutes );
```
2 minutes per page seems quite rapid to me. I don't have any concrete suggestions on what it _should_ be, this just seems quite fast, and maybe someone else will.

Saicchi replied:
Maybe 8 ~ 10 minutes per page? In which case a way to resume the activity is definitely needed.

anothersimulacrum replied:
I'd err towards 10 minutes, but that works.

Cenness replied:
I took some random cooking recipes from a website and tried rewriting them on paper and got following times per pages (A4, 28 lines per page on average):
5:30.42
6:13.90
4:56.59
5:25.45
8:03.70

So median is ~6 minutes, with maximum at 8.

anothersimulacrum replied:
What level skill would they correspond to? I'm caring more about the use case of implied recipes (which cooking recipes definitely are not qualifying as, at least where you're demonstrating), where's it's either a mention you can do something, or a poorly written recipe, or just something you figure out from the information in the book. Given those values, I think this is a reasonable compromise.

anothersimulacrum reviewed (commented):
Looks great! You could consider adding the ability to resume this activity, but that doesn't need to happen here.

I-am-Erk commented:
I think this could be NO_ACTIVITY. You're just sitting and writing

Saicchi commented:
Tested it and didn't find any problems.
Currently the player can copy the recipe even its pitch dark. If I stop the player from continuing copy if it turns dark, his progress will most certainly be reset when he starts another activity like turn on a lamp, unless I do something like save the state of recipe being copied inside the item like EIPC_RECIPES, which I don't know if its something I should do.

anothersimulacrum commented:
Turning on a lamp does not start on activity, so it should not inhibit resumption. Either way, it definitely should be stopping when light runs out.

I-am-Erk commented:
How is it going with this? If you resolve the merge conflicts is it something we can put in?

Saicchi commented:
I haven't checked this in a while, but it should. I'll try.

anothersimulacrum commented on src/iuse.cpp:
```diff
@@ -9794,6 +9786,126 @@ int iuse::magic_8_ball( player *p, item *it, bool, const tripoint & )
+
+    uilist menu;
+    menu.text = _( "Choose recipe to copy" );
+    for( auto r = res.begin(); r != res.end(); r++ ) {
+        recipes.emplace_back( *r );
+        menu.addentry( ( *r )->result_name() );
+    }
```
```suggestion
    for( const recipe &rec : res ) {
        recipes.emplace_back( rec );
        menu.addentry( ( rec )->result_name() );
    }
```

Salty-Panda commented:
Does it allows for storing recipes on electric devices?
I read some issue today and got idea of piggybacking on this PR and adding support for "take a pic of recipe", since we can already store recipes on laptop.

Saicchi commented:
No. But I'm planning on allowing cellphones and eink tablets to take pictures of books to store the recipe and so the player can read the book later.
The code is mostly done, I'm just waiting for my other PR refactoring ACT_READ so I can PR it, but I don't think it will be merged until 0.F.

Salty-Panda commented:
> No. But I'm planning on allowing cellphones and eink tablets to take pictures of books to store the recipe and so the player can read the book later.

Great. So like binder but unlimited and instant?
Do you integrate existing recipes from SD cards?

I understand that you store entire books as fake items inside the laptop/SD. Can NPC read them?

Saicchi commented:
Unlimited but not instant. It won't copy recipes, obly books, but you'll be able to use recipes from the copied books. NPCs will be able to read them.

Saicchi commented:
I've rebased and modernized the code and its ready for a merge. It now properly uses pen/pencil charges.  
Testing revealed nothing unusual. The paper drop and book binder drop rate and numbers might need some work. Hopefully contributors in the future will find a good balanced number for them. The change was necessary due to paper previously being a comestible with charges, and I changed it to an ammo type instead.

The recipe too, was originally done before proficiencies, so I would appreciate if anyone could suggest an improvement to it.

To refresh a bit on how it works:
With a writing tool ( pen, pencil ) and paper the player copies recipes into a book binder. The amount of pages needed is `1 + floor( recipe_difficulty / 2 )`, it takes `10_minutes * pages` minutes to copy a recipe.
Once a recipe is inside, you can't remove it.
The book binder can contain 90 pages.

Recipes in gray don't fit in your book.
![103951](https://user-images.githubusercontent.com/47158232/114562788-345b3080-9c45-11eb-93e8-a9ecdd2a153a.png)

actual-nh commented:
Fix conflicts?

Night-Pryanik commented:
@Saicchi could you please resolve conflicts?

Saicchi commented:
I've rebased to current experimental version. Due to some issues with the leather journal and mre bags ( paper item length was too long at 297mm ), I bumped the leather journal maximum item length to 300mm and for the mre bags I've replaced paper with a new item napkin, which is **exactly** the same as paper before I changed it.

mqrause commented on src/activity_actor.cpp:
```diff
@@ -846,6 +847,79 @@ std::unique_ptr<activity_actor> hacking_activity_actor::deserialize( JsonIn & )
+
+void bookbinder_copy_activity_actor::finish( player_activity &act, Character &p )
+{
+    if( !book_binder->can_contain( item( itype_paper, calendar::turn, pages ) ) ) {
```
```suggestion
    if( !book_binder->can_contain( item( itype_paper, calendar::turn, pages ) ).success() ) {
```

kevingranade reviewed (approved):
Super sorry for the incredibly long delay, every time I had time to look at it, it had been eaten up in conflicts :/

This only needs basic "confirm it works" testing and it'll be good to merge.
I should be able to get to that later today if no one beats me to it.
